src/evaluators/JSXElement.js

```diff
@@ -215,104 +194,113 @@ function evaluateJSXChildren(
+        spreadValue = Environment.GetValue(realm, env.evaluate(astAttribute.argument, strictCode));
 
-        if (spreadValue instanceof ObjectValue) {
-          for (let [key, spreadProp] of spreadValue.properties) {
-            if (spreadProp !== undefined && spreadProp.descriptor !== undefined) {
-              let spreadPropValue = spreadProp.descriptor.value;
-
-              if (spreadPropValue instanceof Value) {
-                if (key === "children") {
-                  children = spreadPropValue;
-                } else {
-                  attributes.set(key, spreadPropValue);
-                }
-              }
+        if (spreadValue instanceof ObjectValue && !spreadValue.isPartialObject()) {
+          for (let [spreadPropKey, binding] of spreadValue.properties) {
+            if (binding && binding.descriptor) {
+              setConfigProperty(spreadPropKey, Get(realm, spreadValue, spreadPropKey));
             }
           }
         } else {
-          throw new FatalError("ObjectValues are the only supported value for JSX Spread Attributes");
+          spreadAttributesCount++;
+          containsAbstractSpreadAttribute = true;
+          invariant(spreadValue instanceof AbstractValue || spreadValue instanceof ObjectValue);
+
+          // we push the props up to this point into the abstract props args. we also
+          // push the abstract spread object and then we create a fresh props object
+          if (objectHasNoPartialKeyAndRef(realm, spreadValue)) {
+            spreadAttributesWithInitialPropsHintCount++;
+            if (config.properties.size > 2) {
+              abstractPropsArgs.push(config);
+            }
+          } else if (config.properties.size > 0) {
+            abstractPropsArgs.push(config);
+          }
+          abstractPropsArgs.push(spreadValue);
+          config = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+          deleteRefAndKeyFromProps(realm, config);
         }
         break;
       default:
-        invariant(false, `Unknown JSX attribute type:: ${astAttribute.type}`);
+        invariant(false, `Unknown JSX attribute type: ${astAttribute.type}`);
     }
   }
-  return {
-    attributes,
-    children,
-  };
-}
 
-function createReactProps(
-  realm: Realm,
-  type: Value,
-  attributes: Map<string, Value>,
-  children,
-  env: LexicalEnvironment
-): ObjectValue {
-  let obj = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
-  for (let [key, value] of attributes) {
-    if (typeof key === "string") {
-      if (RESERVED_PROPS.hasOwnProperty(key)) {
-        continue;
+  if (containsAbstractSpreadAttribute) {
+    // if we haven't assigned any attributes and we are dealing with a single
+    // spread attribute, we can just make the spread object the props
+    if (attributesAssigned === 0) {
+      invariant(spreadValue instanceof ObjectValue || spreadValue instanceof AbstractObjectValue);
+      config = spreadValue;
```

Comment:
I think this isn't safe for the same reason as https://github.com/facebook/prepack/pull/1447#issuecomment-367359080.

This object could have a prototype and could have non-enumerable properties on it. We should filter those out.